### DISCUSSION TITLE
stdin and stdout

### DISCUSSION THREAD

**Discussion by techcatgato** [pybricks/support#1788] [Q&A]
How do I use stdin and stdout so the robot sends the robot a list\str\int\tuple to a non code.pybricks IDE pc and how do I code the pc to receive it? Thanks in advance

**Comment by dlech** [comment]
When using the communication described in https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/, everything sent over the air has to get converted to bytes. There are a number of ways this can be done. `repr` and `eval`, `struct.pack` and `struct.unpack` (module is `ustruct` instead of `struct` in Pybricks MicroPython), probably some others I'm not recalling at the moment

**Reply by techcatgato** [reply]
Yeah but what are the lines used to send and receive this data?

**Reply by dlech** [reply]
On the hub:

`stdout.buffer.write(b"rdy")` sends data from hub to PC

`cmd = stdin.buffer.read(3)` reads data from PC to hub

On the PC:

`await send(b"fwd")` sends data from PC to hub

`def handle_rx(_, data: bytearray):` receives data from hub to the PC

**Reply by techcatgato** [reply]
on stdout.buffer.write() I have to use bytes right? I am trying to convert ints\tuples\strings to bytes but it always returns None:
```
packedx = ustruct.pack("i", 654)
unpacked = ustruct.unpack("i", packedx)
print(packedx)
```
returns None

**Reply by dlech** [reply]
This is working for me (on SPIKE Prime):

```python
import ustruct

packedx = ustruct.pack("<i", 654)
unpacked, = ustruct.unpack("<i", packedx)
print(packedx, unpacked)
```

Prints:

```console
b'\x8e\x02\x00\x00' 654
```

A couple of tips:
* I added `<` to specify endianness. Most computers are little endian, so probably doesn't make much difference. But since data is going between computers, it's always nice to have an explicit endianness.
* `unpack()` returns a tuple, so the extra `,` after `unpacked` unpacks the tuple.

**Reply by DylanBPY** [reply]
> On the hub:
> 
> `stdout.buffer.write(b"rdy")` sends data from hub to PC
> 
> `cmd = stdin.buffer.read(3)` reads data from PC to hub
> 
> On the PC:
> 
> `await send(b"fwd")` sends data from PC to hub
> 
> `def handle_rx(_, data: bytearray):` receives data from hub to the PC

Is there a way to have a dynamic amount of bytes in `cmd = stdin.buffer.read(3`?

**Reply by laurensvalk** [reply]
You can check that data is available before reading it, so that it doesn't wait indefinitely when there isn't anything yet: https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/

**Reply by techcatgato** [reply]
> You can check that data is available before reading it, so that it doesn't wait indefinitely when there isn't anything yet: https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/

Not sure if this is what you said or not, But I think he means a way to just send a message without the buffer element, and make the pc read the whole newest message

**Reply by DylanBPY** [reply]
Yes.
